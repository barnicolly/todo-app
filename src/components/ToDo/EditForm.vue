<template>
  <form @submit.prevent="onSubmit">
    <div>
      <input
        :id="id"
        type="text"
        autocomplete="off"
        v-model.lazy.trim="newTitle" />
    </div>
    <div>
      <button @click="onSubmit" type="button">Сохранить</button>
      <button type="button" @click="onCancel">Отмена</button>
    </div>
  </form>
</template>

<script setup>
const props = defineProps({
  title: {
    type: String,
    required: true,
  },
  id: {
    type: String,
    required: true,
  },
});

const newTitle = defineModel('newTitle');
newTitle.value = props.title;

const emit = defineEmits(['item-edited', 'edit-cancelled']);

function onSubmit() {
  if (newTitle.value && newTitle.value !== props.title) {
    emit('item-edited', newTitle.value);
  }
}

function onCancel() {
  emit('edit-cancelled');
}
</script>
