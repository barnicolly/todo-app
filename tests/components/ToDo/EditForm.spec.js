import { shallowMount } from '@vue/test-utils';
import { faker } from '@faker-js/faker';
import EditForm from '@/components/ToDo/EditForm.vue';

// todo-misha переделать селекторы как будут классы;
describe('EditForm.vue', () => {
  let wrapper;
  let id;
  let title;

  beforeEach(() => {
    id = faker.string.uuid();
    title = faker.lorem.sentence();
    wrapper = shallowMount(EditForm, {
      props: {
        title,
        id,
      },
    });
  });

  it('should correct initialize', () => {
    const input = wrapper.find('input');
    expect(wrapper.vm.newTitle).toBe(title);

    expect(input.attributes().id).toBe(id);
    expect(input.element.value).toContain(title);
  });

  it('click cancel btn', () => {
    wrapper.find('button:nth-child(2)').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('edit-cancelled');
  });

  it('click save btn', async () => {
    const input = wrapper.find('input');

    input.element.value = 'new value';
    await input.trigger('change');

    expect(wrapper.vm.newTitle).toBe('new value');
    wrapper.find('button:nth-child(1)').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('item-edited');
  });
});

describe('EditForm.vue Snapshot', () => {
  it('has same HTML structure', () => {
    const wrapper = shallowMount(EditForm, {
      props: {
        title: 'some-title',
        id: 'some-uuid',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
