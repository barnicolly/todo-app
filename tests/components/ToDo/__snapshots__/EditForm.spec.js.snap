// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditForm.vue Snapshot has same HTML structure 1`] = `
<form>
  <div><input id="some-uuid" type="text" autocomplete="off"></div>
  <div><button type="submit"> Сохранить </button><button type="button"> Отмена </button></div>
</form>
`;
